deploy/kubernetes/elastic-agent-managed-kubernetes.yaml

```diff
@@ -64,6 +64,14 @@ spec:
                   fieldPath: metadata.name
           securityContext:
             runAsUser: 0
+            capabilities:
+              drop:
+                - all
+              add:
+              # Needed for 'Defend for containers' integration (cloud-defend)
+                - BPF
```

Comment:
Adding these `capabilities` into the manifests for the generic use case looks concerning to me. I would propose these specifically to be commented out and only being enabled when users actually want to.
We also need to document what behavior these capabilities enable and their scope.

Reply:
Have you tested that this setting does not cause any issue when a user has enabled just Kubernetes and system integration?

Reply:
@ChrsMark, I am more than glad to document these `capabilities`. What is the expectation? Should I describe the scope of each cap in a comment directly in the YAML, or is there more extensive documentation you are referring to? 

Let me know, and I will make it happen.

@MichaelKatsoulis I have not tested this setting with the Kubernetes and system integrations. Can you point me to documentation or describe an appropriate set of tests to validate that this does not break things? Do you have a suggested smoke test? 

Finally, I have concerns about commenting out these `capabilities` in the code. It will mean that users cannot discover our new integration without it starting in a broken state. I have made significant efforts to reduce this set of capabilities to the minimum required for our product to work. I do consider this a stop-gap solution until we can leverage Kustomize and would like to assist in those efforts. 

@learhy can you provide additional context as to the product requirements around discoverability?

Reply:
Hey @norrietaylor from my side that would help. I would add those info as comments in the manifests and maybe inside https://www.elastic.co/guide/en/fleet/master/running-on-kubernetes-managed-by-fleet.html.

Keep in mind that we have features like Hints' autodiscovery that are also disabled by default and their parts are commented out in the manifests. For example: https://github.com/elastic/elastic-agent/blob/main/deploy/kubernetes/elastic-agent-standalone-kubernetes.yaml#L28-L29.

Last but not least I see the need for product discoverability however I'm also concerned about how we evaluate these additions specially when those are mostly security permissions. @gizas @ruflin @cmacknz I would like your input here. Shall we establish a policy or specific criteria around this? Is it ok that by default Agent on k8s comes with a superset of permissions and accesses? 
As a devops/SRE user I would be totally fine if I were told to manually enable advanced features while I would be concerned if everything is by default open in anycase.

Reply:
For me the superset approach is not optimal. As long as we dont use the `least privilege approach` we are prone to errors and misconfigurations to users.

The thing is that I dont have a new solution to suggest. Am I thinking loud here that a second manifest here for security team is a good solution? (But then how we keep them in sync)
Kustomize can be the solution here once we provide the first draft. But again we need to make sure how to run the full tests when kustomize will produce the manifests

So I note here two problems the housekeeping of different version of manifests and then the testing of different versions.

Reply:
I am in total agreement that the superset approach is not optimal. Unfortunately, the current architecture dictates this model. By placing all software required to run all integrations in the same container, we cannot escape violation of the principle of least privilege.

It isn't just a problem for `capabilities,` but also mount points, k8s permissions and more. Some of the new work we are planning for Kustomize and a potential Fleet operator will likely allow us to solve this problem; however, we have already architecturally committed to this model by having a single Agent image and daemonest. This request isn't the first time a new integration has had to increase the scope of this YAML, and I am sure it won't be the last until we have a more cloud-native approach to deploying software in Kubernetes.

I would also like to call out the fact that I have worked in the patch to drop all existing and unaudited `capabilities` to increase the security posture of this container.

Reply:
From a product perspective, the UX of starting an agent and then having it fail right away isn't a good look. It’d look bad to our sales engineering team, and to any customer that tries it out on their own.  If we are lucky, they might come back to find the documentation, but only if they are already convinced of our value. Prospective customers that are just getting initial impressions (or are not very familiar with Kubernetes) will find the agent not working, assume we released a bug, and-- likely never give us another look.

As an aside, it actually feels like we've made an improvement here by dropping all caps...

Reply:
> I am in total agreement that the superset approach is not optimal. Unfortunately, the current architecture dictates this model. By placing all software required to run all integrations in the same container, we cannot escape violation of the principle of least privilege.

The underlying Elastic Agent container today bundles everything together, but there isn't really a reason why we can't have multiple reference configurations for deploying on Kubernetes.

Would it help to separate the YAML here directly into the two separate use case of "I want to get started with Kubernetes observability" and "I want to get started with container protection"? 

At minimum, separating the use cases makes it far less likely a change from the team maintaining one of these initial configurations unintentionally breaks something for the other one. 

I would rather us maintain multiple configurations with some duplication that are focused and well tested for their core use case, than have one monolithic configuration that breaks regularly because it needs to be tested across a large combination of use cases.

There is already a separate set of Kubernetes YAML for getting started with endpoint in for example as a precedent: https://github.com/elastic/endpoint/blob/main/releases/8.6.0/kubernetes/deploy/elastic-defend.yaml

Reply:
> There is already a separate set of Kubernetes YAML for getting started with endpoint in for example as a precedent: https://github.com/elastic/endpoint/blob/main/releases/8.6.0/kubernetes/deploy/elastic-defend.yaml

@cmacknz this YAML is being deprecated for precisely the reason that @learhy mentioned above.  Also, if a user wants to add o11y and security integrations how do they juggle the YAML? 

> Would it help to separate the YAML here directly into the two separate use case of "I want to get started with Kubernetes observability" and "I want to get started with container protection"?

I don't think so. I would expect that the same users want to use the k8s o11y tooks as the one we are adding here. They probably also would like to use KSPM, which already has permissions and mounts added to this YAML.

There is prior art here as well. When extra permissions for cloudbeat were needed they were added into this YAML. Why can't we follow this model while we work toward an operator? 

> I would rather us maintain multiple configurations with some duplication that are focused and well tested for their core use case, than have one monolithic configuration that breaks regularly because it needs to be tested across a large combination of use cases.

I just want to add a few capabilities. This seems like a minor change and should not break other solutions. I am glad to remove the `drop all` line as that is the only thing that introduces risk for the o11y team. My intention here was only to improve the ser

Reply:
@cmacknz, @ChrsMark. 

I've updated the YAML with documentation and have dropped the potentially dangerous `drop all` statement. In addition, I have added some language to inform the customer that these caps and mounts can be removed if the cloud-defend integration is not being used. This adheres to what previous teams have done and is consistent in the YAML.